x-pack/plugin/identity-provider/src/main/java/org/elasticsearch/xpack/idp/privileges/UserPrivilegeResolver.java

```diff
@@ -0,0 +1,110 @@
+import org.elasticsearch.xpack.core.security.authz.RoleDescriptor;
+import org.elasticsearch.xpack.core.security.authz.permission.ResourcePrivileges;
+
+import java.util.HashSet;
+import java.util.Map;
+import java.util.Optional;
+import java.util.Set;
+import java.util.stream.Collectors;
+
+/**
+ * Determines what privileges a user has within a given {@link ServiceProviderPrivileges service}.
+ */
+public class UserPrivilegeResolver {
+
+    public class UserPrivileges {
+        public final String principal;
+        public final boolean hasAccess;
+        public final Set<String> groups;
+
+        public UserPrivileges(String principal, boolean hasAccess, Set<String> groups) {
+            this.principal = principal;
+            this.hasAccess = hasAccess;
+            this.groups = groups;
+        }
+
+        @Override
+        public String toString() {
+            return getClass().getSimpleName() + "{" + principal + ", " + hasAccess + ", " + groups + "}";
+        }
+    }
+
+    private final Logger logger = LogManager.getLogger();
+    private final Client client;
+    private final SecurityContext securityContext;
+
+    public UserPrivilegeResolver(Client client, SecurityContext securityContext) {
+        this.client = client;
+        this.securityContext = securityContext;
+    }
+
+    /**
+     * Resolves the user's privileges for the specified service.
+     * Requires that the active user is set in the {@link org.elasticsearch.xpack.core.security.SecurityContext}.
+     */
+    public void resolve(ServiceProviderPrivileges service, ActionListener<UserPrivileges> listener) {
+        HasPrivilegesRequest request = new HasPrivilegesRequest();
+        final String username = securityContext.getUser().principal();
+        request.username(username);
+        request.applicationPrivileges(buildResourcePrivilege(service));
+        client.execute(HasPrivilegesAction.INSTANCE, request, ActionListener.wrap(
+            response -> {
+                logger.debug("Checking access for user [{}] to application [{}] resource [{}]",
+                    username, service.getApplicationName(), service.getResource());
+                UserPrivileges privileges = buildResult(response, service);
+                logger.debug("Resolved service privileges [{}]", privileges);
+                listener.onResponse(privileges);
+            },
+            listener::onFailure
+        ));
+    }
+
+    private UserPrivileges buildResult(HasPrivilegesResponse response, ServiceProviderPrivileges service) {
+        final Set<ResourcePrivileges> appPrivileges = response.getApplicationPrivileges().get(service.getApplicationName());
+        if (appPrivileges == null || appPrivileges.isEmpty()) {
+            return new UserPrivileges(response.getUsername(), false, Set.of());
+        }
+        final boolean hasAccess = checkAccess(appPrivileges, service.getLoginAction(), service.getResource());
```

Comment:
Wondering if we can short-circuit here and not check for group actions. If the user can't sso then the groups will never be "released" as attributes to the SP so why check them and not return an empty list ? Do we foresee exposing this externally where such answers "Can't login, but if they could they would be a `viewer` " would be useful?

Reply:
We can.
To my thinking these were separate checks so the more accurate return value is to populate the groups even if `hasAccess` is false, but we can document it so that `groups` will ways be empty is `hasAccess` is false.
I think that's a reasonable behaviour.

Reply:
I think the _logs_ will need to be super clear though. I can see people getting confused because the logs say `groups=[]` and they try to chase down that "problem" when in fact that's _not_ the relevant piece of information.

Reply:
I agree that we need to be explicit and somewhat descriptive in the logs eitherway. Even if we did return the group actions we would need to be explicit that the issue is with the login action otherwise the user/admin might think that they just miss the "necessary" group